testfiles/catalog-local.json

```diff
@@ -0,0 +1,13 @@
+{
+  "version": 1.0,
+  "schemas": [
+    {
+      "name": "Valid custom schema",
+      "fileMatch": [
+        "valid.json",
+        "valid.yml"
+      ],
+      "url": "./schema.json"
```

Comment:
So I hadn't really thought about this until I saw this code, but this opens a bit of a can of worms.. I'll try and summarise my thoughts.

- This `url` is not a URL - its a local path, which is confusing
- The schema catalog schema says a schema URL is

  ```json
  "url": {
    "description": "An absolute URL to the schema location",
    "type": "string",
    "format": "uri",
    "pattern": "^https://"
  }
  ```
  so this is now not strictly a schema catalog.
- If we ignore this and allow `url` to be a path to a local file this introduces an important question: if the path to the schema is relative, what is it relative to?
  I think in your original implementation you're saying the path of the schema is relative to the path of the catalog file?
  In the one I've suggested above the path is assumed to be relative to the execution context (i.e: we just treat it the same as an arg passed to `--schema`), so this would become `./testfiles/schema.json`
  That's kind of wierd, but the concept of the schema path being relative to the catlog path gets even wierder if the catalog is a remote URL.

I'm undecided on where to go with this.. I think there are two options worth considering, neither of which are completely satisfying:
- Only allow `schema.url` be be an actual URL.
- Break with the schema catalog definition, allow `schema.url` to also be a path (absolute or relative). If the path is relative interpret it relative to the execution context.

The first is more technically pure. A user-supplied schema catalog has the same rules as schemastore. However it means we can specify either a remote or local schema with `--schema` but there is no way to specify we want to use a local schema with `--catalogs`
The second is annoying because `schema.url` might be.. not a URL. We're inventing our own thing rather than leveraging an existing format and your catalog might contain relative paths that don't make sense relative to the catalog itself, depending on where you put it. It does give us a way to address the local schema use-case though. I'm somewhat torn..

Interested in your thoughts on the tradeoffs here as someone who will presumably actually use this feature...

Reply:
For now, I have implemented a solution that manage local schemas relative to the execution context despite the expected catalog format in dad46a136bc280a1863005a941aa9363e03e0a28.

It also allow a strange behavior (but that could make sense to someone ??). Take a look at the "should return 0 when file is valid (with auto-detected schema from custom remote catalog matching relativly defined schema)" test.

Reply:
> This I why I remove the $schema property in test files.

Ah yes. I had not spotted this.

> I was first thinking about adding a `path` property to match against local schemas

Yes this is another option I considered. Again, it is not completely satisfying but on reflection maybe better then abusing `url` if we need to diverge anyway.

> I do not have a need for using local schemas in a local catalog

OK. This is a useful data point.
I know this is becoming a bit annoying to work on as I am changing the requirements as we go along, but I think a "safe" starting point for this feature is we say `schema.url` can only be a URL. This is unambiguously correct wherever the catalog lives. It makes it easy to understand/explain the feature and it solves the problem you (and some other users) are having. Doing this also doesn't prohibit us (me) from adding the ability to use a local schema (in some way or another - implementation to be decided) at a later date.

Essentially what worries me is that all the solutions we've come up with so far between the two of us for specifing local schemas are a bit dubious in some way - we haven't really hit on anything that is obviously correct or intuitive. What concerns me as a maintainer is that I merge a solution which is somewhat half-baked, people start relying on it and then it becomes hard to change. Of course, the project is pre-1.0 so I can make breaking changes at will, but I would rather not do this :) It may also be easier to understand what the right solution should be once we have real examples of users who want to create a catalog referencing local schemas (or maybe they never materialize and it becomes a non-issue).

Lets start with:

- catalog can be a remote URL or a local path
- `schema.url` is always a URL

We can get the feature landed on that basis and worry about extending it as another issue.